src/Collections/Sequences/Sorting.dfy

```diff
@@ -0,0 +1,173 @@
+      }
+    }
+    assert AntiSymmetric(LexicographicByteSeqBelow) by {
+      forall x, y, n: nat |
+        n <= |x| && n <= |y| && x[..n] == y[..n] &&
+        LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, x, n)
+      {
+        AboutLexicographicByteSeqBelowAux_AntiSymmetric(x, y, n);
+      }
+    }
+    assert StandardLibrary.Transitive(LexicographicByteSeqBelow) by {
+      forall x, y, z, n: nat |
+        n <= |x| && n <= |y| && n <= |z| &&
+        LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, z, n)
+      {
+        AboutLexicographicByteSeqBelowAux_Transitive(x, y, z, n);
+      }
+    }
+    assert Connected(LexicographicByteSeqBelow) by {
+      forall x, y, n: nat | n <= |x| && n <= |y| {
+        AboutLexicographicByteSeqBelowAux_Connected(x, y, n);
+      }
+    }
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Reflexive(x: seq<uint8>, n: nat)
+    requires n <= |x|
+    ensures LexicographicByteSeqBelowAux(x, x, n)
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_AntiSymmetric(x: seq<uint8>, y: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y|
+    requires x[..n] == y[..n]
+    requires LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, x, n)
+    ensures x == y
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Transitive(x: seq<uint8>, y: seq<uint8>, z: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y| && n <= |z|
+    requires LexicographicByteSeqBelowAux(x, y, n) && LexicographicByteSeqBelowAux(y, z, n)
+    ensures LexicographicByteSeqBelowAux(x, z, n)
+    decreases |x| - n
+  {
+  }
+
+  lemma AboutLexicographicByteSeqBelowAux_Connected(x: seq<uint8>, y: seq<uint8>, n: nat)
+    requires n <= |x| && n <= |y|
+    ensures LexicographicByteSeqBelowAux(x, y, n) || LexicographicByteSeqBelowAux(y, x, n)
+    decreases |x| - n
+  {
+  }
+
+  predicate method SortedBy<T>(a: seq<T>,compare: (T, T) -> bool ) {
+    forall i, j | 0 <= i < j < |a| :: compare(a[i],a[j])
+  }
+
+  lemma LemmaNewFirstElementStillSortedBy<T>(x: T, s: seq<T>,  compare: (T, T) -> bool) 
+    requires SortedBy(s, compare)
+    requires |s| == 0 || compare(x,s[0])
+  {}
+
+function method MergeSortBy<T>(a: seq<T>, compare: (T, T) -> bool ): (result :seq<T>)
+    ensures multiset(a) == multiset(result)
+  {
+    if |a| <= 1 then
+      a
+    else
+      var splitIndex := |a| / 2;
+      var left, right := a[..splitIndex], a[splitIndex..];
+
+      assert a == left + right;
+
+      var leftSorted := MergeSortBy(left, compare);
+      var rightSorted := MergeSortBy(right, compare);
+      
+      MergeSortedBy(leftSorted, rightSorted, compare)
+  }
+
+ function method {:tailrecursion} MergeSortedBy<T>(left: seq<T>, right: seq<T>,  compare: (T, T) -> bool) : (result :seq<T>)
+    requires SortedBy(left,compare)
+    requires SortedBy(right,compare)
+    ensures multiset(left + right) == multiset(result)
```

Comment:
This result should also be sorted.

Reply:
I think this method will also need the same additional preconditions that `MergeSortBy` does.